### PR TITLE
Introduce ambient_affine_space of an affine scheme

### PR DESCRIPTION
**Pull request by simonbrandhorst**
Following our discussion, this introduces the `ambient_affine_space` and the `ambient_coordinate_ring` and
`ambient_closure_ideal` (I am not really happy with `ambient_closure_ideal` ... so if anyone has a better name. Please come forward.  But in the end I think that most of these things are best addressed by going through the corresponding morphisms. 
- I also changed `ambient_ring` of a projective scheme to `ambient_coordinate_ring` (`ambient_homogeneous_coordinate_ring`) seemed too long. 
- I excluded the copy constructor from the official documentation, because I have never seen this type of "copy constructor" in julia. It seems like a java thing. (So this needs further discussion before it becomes part of the "official" part of Oscar.

### CODE REVIEW

File: docs/src/AlgebraicGeometry/Schemes/AffineSchemes.md [hunk 1]
```diff
@@ -75,26 +69,23 @@ closure(X::AbsSpec, Y::AbsSpec)
...
 ```@docs
-ambient_ring(X::AbsSpec)
+ambient_affine_space(X::AbsSpec)
 ```
-For instance, this is the case whenever ``R`` is a quotient
-ring of ``P``, a localization of ``P``, or a localization
-of a quotient ring of ``P``; but also for
-power series rings in multiple variables.
+Some such as ``Spec(ZZ)`` or ``Spec(k)`` for ``\mathbb k``
+a field do not have an ambient affine space.
```
Comment (afkafkafk13) [inline docs/src/AlgebraicGeometry/Schemes/AffineSchemes.md]: This is not precise enough for documentation: 
What is the return, if X=Spec(R) is not embedded? Does it throw an error exception or does it return a sensible default value, which can be checked, like e.g. X === ambient_affine_space(X) or A^0_R?

File: docs/src/AlgebraicGeometry/Schemes/ArchitectureOfAffineSchemes.md [hunk 2]
```diff
@@ -66,8 +67,8 @@ These must return representatives of the numerator and the denominator
 of ``a``. Note that the denominator is equal to `one(P)` in case
 ``R \cong P`` or ``R \cong P/I``.
 
-Given the above characterization of the `ambient_ring(X)`,
-we shall refer to the ``x_i`` above as the *coordinates* of ``X``.
+Recall that the coordinates ``x_i`` of ``X`` are induced by the coordinates of
+the ambient space.
```
Comment (afkafkafk13) [inline docs/src/AlgebraicGeometry/Schemes/ArchitectureOfAffineSchemes.md]: ambient **affine** space  (to stay in the wording of the commands and documentation)

**Review by HechtiDerLachs — APPROVED** [review]

File: src/Rings/mpoly-localizations.jl [hunk 3]
```diff
@@ -922,6 +922,7 @@ inverted_set(W::MPolyLocalizedRing) = W.S
 
 ### additional getter functions
 gens(W::MPolyLocalizedRing) = W.(gens(base_ring(W)))
+ngens(W) = ngens(base_ring(W))
```
Comment (afkafkafk13) [inline src/Rings/mpoly-localizations.jl]: Are you sure that you do not want to specify a type for W? What would happen, if W were an MPolyRing over k and base_ring(W) returns k?

**Comment by HereAround** [comment]
It might be good to mark the old methods as deprecated. This way, the old method names are still supported for those users that are already using them. And they get informed that the names have changed.

File: src/Schemes/AffineSchemes/Objects/Attributes.jl [hunk 4]
```diff
@@ -22,38 +23,223 @@ On an affine scheme ``X = Spec(R)`` this returns the ring ``R``.
...
+julia> V = hypersurface_complement(Y, y)
+
+julia> ambient_affine_space(U) === X
+true
+```
+
+We can create ``X``, ``Y`` and ``Z`` also by first constructing the corresponding
+coordinate rings. The subset relations are inferred from the coordinate rings.
+More precisely, for a polynomial ring ``P`` an ideal ``I < P `` and a multiplicatively closed subset
+``U`` of ``P`` let ``R`` be one of ``P``, ``P/I`` or ``U^{-1}(P/I)``.
```
Comment (afkafkafk13) [inline src/Schemes/AffineSchemes/Objects/Attributes.jl]: U^{-1}P is missing in the list

**Comment by simonbrandhorst** [comment]
It was on experimental. By the definition of experimental, the code is unstable. So I wouldn't bother with deprecation. 
And the part that is not on experimental shouldn't be part of a release yet. (I just left the old aliases, so that we do not run into merge conflicts with the cleanup PRs. But if the interface for something exported in src/ changes in future, we would indeed have to deprecate.

**Comment by wdecker** [comment]
What is an ambient closure ideal? Let us please select names which are at least close to self-explaining.

**Comment by simonbrandhorst** [comment]
> What is an ambient closure ideal? Let us please select names which are at least close to self-explaining.

For $U \subseteq X \subseteq \mathbb{A}^n$ this is the ideal of $X$, i.e. the defining ideal of the closure of $U$ in $\mathbb{A}^n$. I could live with not exporting this function at all.

File: src/Schemes/AffineSchemes/Objects/Attributes.jl [hunk 5]
```diff
@@ -110,10 +304,7 @@ end
 @Markdown.doc """
     codim(X::Spec)
 
-In Oscar, we can compute for an affine scheme ``X``
-the ring ``R = ambient_ring(X)``. This allows to consider
-an embedding of ``X`` into ``Spec(R)``. This method returns
-the codimension of the image of this embedding in ``Spec(R)``.
+Return the codimension of ``X`` in its ambient affine space.
```
Comment (afkafkafk13) [inline src/Schemes/AffineSchemes/Objects/Attributes.jl:329]: Maybe,  document that it asserts that there is an ambient affine space.

File: src/Schemes/AffineSchemes/Objects/Attributes.jl [hunk 6]
```diff
@@ -202,72 +396,32 @@ julia> (x1,x2,x3) = gens(R)
 julia> Y = subscheme(X,ideal(R,[x1*x2]))
 Spec of Quotient of Multivariate Polynomial Ring in x1, x2, x3 over Rational Field by ideal(x1*x2)
 
-julia> defining_ideal(Y)
+julia> ambient_closure_ideal(Y)
```
Comment (afkafkafk13) [inline src/Schemes/AffineSchemes/Objects/Attributes.jl]: For better understanding of the setting, maybe add another line for the base_ring of this ideal.

File: src/Schemes/AffineSchemes/Objects/Attributes.jl [hunk 7]
```diff
@@ -202,72 +396,32 @@ julia> (x1,x2,x3) = gens(R)
...
-
-########################################################################
-# (3) Further attributes for Spec
-########################################################################
-
-strict_modulus(X::Spec) = saturated_ideal(modulus(OO(X)))
-
+@attr ambient_closure_ideal(X::AbsSpec{<:Any, <:MPolyRing}) = ideal(OO(X), [zero(OO(X))])
+ambient_closure_ideal(X::AbsSpec{<:Any, <:MPolyQuo}) = modulus(OO(X))
+@attr ambient_closure_ideal(X::AbsSpec{<:Any, <:MPolyLocalizedRing}) = ideal(OO(X), [zero(OO(X))])
```
Comment (afkafkafk13) [inline src/Schemes/AffineSchemes/Objects/Attributes.jl]: If it is the closure in the ambient affine space, then this lives in OO(ambient_affine_space) by naming convention and not in OO(X) in the localized setting. If this is to live in the localized ring (a setting for which such a command is needed as well !!), then we need a different name respecting the localization (which defaults to ambient_closure_ideal for non-localized rings).

**Review by afkafkafk13 — COMMENTED** [review]
I am not yet at ease with ambient_closure_ideal, but I do not have a better name for it.

There is also a small question about how to handle non-embedded spectra, if handed accidentally to a method or how to properly detect them to avoid handing them to a method. Maybe Is_embedded?

**Comment by HechtiDerLachs** [comment]
> There is also a small question about how to handle non-embedded spectra, if handed accidentally to a method or how to properly detect them to avoid handing them to a method. Maybe Is_embedded?

As I already suggested yesterday: I think this only makes sense for spectra of `MPolyAnyRing`s and should simply be documented that way. So if you want to make sure not to accidentally write a method for something that does not have `ambient_affine_space`, then make that part of your method's signature.

File: docs/src/AlgebraicGeometry/Schemes/AffineSchemes.md (hunk 1 above)
Comment (HechtiDerLachs) [reply docs/src/AlgebraicGeometry/Schemes/AffineSchemes.md]: See my comment below.

File: src/Schemes/AffineSchemes/Objects/Attributes.jl (hunk 6 above)
Comment (HechtiDerLachs) [reply src/Schemes/AffineSchemes/Objects/Attributes.jl]: I don't understand what you mean here.

File: src/Schemes/AffineSchemes/Objects/Attributes.jl (hunk 7 above)
Comment (HechtiDerLachs) [reply src/Schemes/AffineSchemes/Objects/Attributes.jl]: You are right. Here the previous implementation differs from the needs you formulated yesterday. I think this is up to you how to resolve this. Unless it breaks the code, of course.

File: src/Schemes/AffineSchemes/Objects/Attributes.jl (hunk 6 above)
Comment (afkafkafk13) [reply src/Schemes/AffineSchemes/Objects/Attributes.jl]: We are in the documentation here. the returned ideal lives in R. Maybe stress this by adding a line checking that the returned ideal is an ideal in R -- for clarifying to the user, where the object lives.

File: src/Schemes/AffineSchemes/Objects/Attributes.jl (hunk 7 above)
Comment (afkafkafk13) [reply src/Schemes/AffineSchemes/Objects/Attributes.jl]: From my perspective, the only consistent way to do this is:
- ambient_... always refers to objects related to ambient_affine_space
- ambient_closure_ideal then has to provide an ideal in OO(ambient_affine_space)
- is there any good name for the localized equivalent or do we expect that users to use modulus(OO(X)) for that?

**Comment by afkafkafk13** [comment]
> > There is also a small question about how to handle non-embedded spectra, if handed accidentally to a method or how to properly detect them to avoid handing them to a method. Maybe Is_embedded?
> 
> As I already suggested yesterday: I think this only makes sense for spectra of `MPolyAnyRing`s and should simply be documented that way. So if you want to make sure not to accidentally write a method for something that does not have `ambient_affine_space`, then make that part of your method's signature.

Yes, this makes sense.

**Comment by simonbrandhorst** [comment]
For me `kernel(pullback(ambient_embedding(X)))` would be the most natural way to get the `ambient_closure_ideal(X)`.  We could promote that in the documentation.
I am not convinced of providing a complicated to remember 
shortcut. It makes the code harder to read for everyone who forgot/does not know what an `ambient_closure_ideal` is.
 `defining_ideal_of_closure_in_ambient_affine_space` is  clean, but too long to type.

File: docs/src/AlgebraicGeometry/Schemes/AffineSchemes.md (hunk 1 above)
Comment (afkafkafk13) [reply docs/src/AlgebraicGeometry/Schemes/AffineSchemes.md]: This is clear now and exactly along the lines of @HechtiDerLachs comment.

**Comment by afkafkafk13** [comment]
> For me `kernel(pullback(ambient_embedding(X)))` would be the most natural way to get the `ambient_closure_ideal(X)`. We could promote that in the documentation. 
Yes, this one looks sufficiently close to an algebraic geometer's notation and sufficiently easy to remember so that we can promote in the documentation. 

> I am not convinced of providing a complicated to remember shortcut. It makes the code harder to read for everyone who forgot/does not know what an `ambient_closure_ideal` is. `defining_ideal_of_closure_in_ambient_affine_space` is clean, but too long to type.
I agree for the case of a difficult ot remember shortcut name, as long as he task is actually covered in the documentation in as much length as we would give the command (i.e. providing a short explicit example as a kind of recipe). 
But we should keep our eyes open whether over time scripting users tend to define their own shortcut for this. If so, we should again think about adding the shortcut with an intuitive name (which might at that point come from the community already using shortcuts).

**Review by afkafkafk13 — APPROVED** [review]

File: src/Schemes/AffineSchemes/Objects/Attributes.jl (hunk 7 above)
Comment (HechtiDerLachs) [reply src/Schemes/AffineSchemes/Objects/Attributes.jl]: > is there any good name for the localized equivalent or do we expect that users to use modulus(OO(X)) for that?

Yes, why not? From what I remember, this was intended as a scheme-sided analogue of `modulus(OO(X))`. But I don't remember the necessity to give this a shortcut name. So I think `modulus(OO(X))` is just as fine as it is not even longer to write.

Comment (afkafkafk13) [reply src/Schemes/AffineSchemes/Objects/Attributes.jl]: I would now vote for @simonbrandhorst's suggestion to completely omit this ambient_closure_ideal as a command, but document well in the scheme documentation how to get to such ideals (by the kernel(pullback(....)) variant for the inclusion in the ambient_affine_space and by the modulus(OO(X)) for a 'localized variant'.

Comment (HechtiDerLachs) [reply src/Schemes/AffineSchemes/Objects/Attributes.jl]: The only downside of this is that going through all these commands, one can not make use of previously cached stuff. Internally, this ideal sits there as a `saturated_ideal`. And in principle, one would be computing this saturation twice. But the knowledgeable user will also know about this way to get their hands on it and if in doubt, one has to program a dispatch for every type of ring, anyway. 

Long story short: If you don't advocate for keeping `ambient_closure_ideal`, we can remove it. Unless this breaks our code, of course.

Comment (afkafkafk13) [reply src/Schemes/AffineSchemes/Objects/Attributes.jl]: Fine with me, as long as the users' need for getting their hands on these ideals is satisfied by suitable documentation.

**Comment by simonbrandhorst** [comment]
I am not sure why the g-module stuff fails. But in any case I do not have the time today to get this PR trough. 
If you need it to proceed, feel free to do the little changes as discussed and get it merged.

File: src/Schemes/AffineSchemes/Objects/Attributes.jl (hunk 7 above)
Comment (simonbrandhorst) [reply src/Schemes/AffineSchemes/Objects/Attributes.jl]: I keep it for now, but I do not export it.

Comment (HechtiDerLachs) [reply src/Schemes/AffineSchemes/Objects/Attributes.jl]: > I keep it for now, but I do not export it.

I think that's reasonable for the time being. We should keep an eye on it as we progress. Because, from my experience, you get a clearer picture over time on how exactly you want and need to use things. If it proves to be really necessary in the long run, we will then eventually know better why and for what.

**Comment by simonbrandhorst** [comment]
@fingolfin  I tried to make the docs build for an hour, but apparently I am missing something. Could you please help?
In particular, referencing methods does not work for me.

**Comment by afkafkafk13** [comment]
@simonbrandhorst @fingolfin : documentation now passes, but I broke a test on the merge from upstream. I am on it....

**Comment by afkafkafk13** [comment]
@HechtiDerLachs @HereAround @simonbrandhorst : I found the problem of the Glueing tests -- it is completely unrelated to this PR. It is a forgotten change "ambient" to "ambient_scheme" from upstream. Is this already fixed upstream or do you want to fix it upstream or do you want me to commit the fix here?

**Comment by HechtiDerLachs** [comment]
> Is this already fixed upstream or do you want to fix it upstream or do you want me to commit the fix here?

I don't really understand what you mean. The tests for the glueings were passing in the main branch, so I guess, they should be fixed here. 

Another thing: We still have #1682 in the pipeline. But this time, we do have some overlap in the `CoveredScheme`s. Nevertheless, I suggest to first merge that PR and then rebase this one on the latest master. I can help with that, then. 

Any objections?

**Comment by afkafkafk13** [comment]
no objections. Just tell me when to close and reopen this one to incorporate the upstream changes.

**Comment by HechtiDerLachs** [comment]
> no objections. Just tell me when to close and reopen this one to incorporate the upstream changes.

No, this is not the point. Any issues with the glueings and their tests should be resolved here and now. There will be no upstream changes. Once #1682 is merged, I will then rebase this branch on the latest master and resolve the issues arising from that update then. No opening and closing required at any point.

**Comment by HereAround** [comment]
General suggestion:

Some of the commit messages indicate that the commit in question/the previous commit might not work (e.g. "try to fix the docs" in https://github.com/oscar-system/Oscar.jl/pull/1683/commits/a9cb784dea755ec31e2c08f44facdbdf3d74447c). As I see it, every commit in the history should work. That way, we can safely switch features on an off as we move back and forth in the git history.

I see two ways to achieve this:
1. Clean the history (squash, rebase, cherry-pick and alike).
2. Squash all changes into a single commit.

Likely, the truth is somewhat intermediate. In particular, if you implement several distinct features, then you likely want them in separate and working commits.

From what I see, here you have a single feature (introduce `ambient_affine_space`), so it feels like squashing all commits might be best.

**Comment by HechtiDerLachs** [comment]
#1682 has been merged. I would now try to rebase this one here on master. That means, I might need to force-push to this branch. Does anyone have uncommitted local changes? If yes, please push them now or tell me to wait!

**Comment by HechtiDerLachs** [comment]
> From what I see, here you have a single feature (introduce `ambient_affine_space`), so it feels like squashing all commits might be best.

I agree. Also, I'm having difficulties with the rebase again, so I would prefer to merge the master instead. This has also already done before, so it shouldn't make things worse. Especially, if we are squashing in the end, anyway.

Would that be OK with everyone?

**Comment by HechtiDerLachs** [comment]
This seemed easier than expected, really. Let's see whether the tests pass.

**Comment by HechtiDerLachs** [comment]
Now tests pass locally for me. Let's see how far we get this time.

**Comment by HechtiDerLachs** [comment]
> In particular, referencing methods does not work for me.

Does this refer to how the docstrings appear in the online documentation? If yes: This is a julia bug and there's nothing we can do about it; see [here](https://github.com/oscar-system/Oscar.jl/issues/1314).

**Comment by simonbrandhorst** [comment]
The point is that the cross reference 
```
See [`inclusion_morphism(::AbsSpec, ::AbsSpec)`](@ref) for a way to obtain the ideal ``I`` from ``X = \mathrm{Spec}(R, I)``. 
```
produces a dead link. ... but perhaps this is also some documenter bug. So the easiest "fix" is to change it to 
```
See `inclusion_morphism(::AbsSpec, ::AbsSpec)` for a way to obtain the ideal ``I`` from ``X = \mathrm{Spec}(R, I)``. 
```
and avoid the links altogether.  And similar for all other references. They do not seem to work.

**Comment by HechtiDerLachs** [comment]
I tried locally without the backticks, but I also get a dead link.

Edit: I get a warning. It seems that we do not make correct use of the `@ref`:
```
┌ Warning: invalid local link: unresolved path in AlgebraicGeometry/Schemes/AffineSchemes.md
│   link.text =
│    1-element Vector{Any}:
│     "inclusion_morphism(::AbsSpec, ::AbsSpec)"
│   link.url = "@ref"
└ @ Documenter.Writers.HTMLWriter ~/.julia/packages/Documenter/yf96B/src/Writers/HTMLWriter.jl:2081
```

**Comment by afkafkafk13** [comment]
> I tried locally without the backticks, but I also get a dead link.
> 
> Edit: I get a warning. It seems that we do not make correct use of the `@ref`:
> 
> ```
> ┌ Warning: invalid local link: unresolved path in AlgebraicGeometry/Schemes/AffineSchemes.md
> │   link.text =
> │    1-element Vector{Any}:
> │     "inclusion_morphism(::AbsSpec, ::AbsSpec)"
> │   link.url = "@ref"
> └ @ Documenter.Writers.HTMLWriter ~/.julia/packages/Documenter/yf96B/src/Writers/HTMLWriter.jl:2081
> ```

You get that warning, when you do not put in the backticks -- and the warning causes a doctest-error during the CI-tests.

File: docs/src/AlgebraicGeometry/Schemes/AffineSchemes.md [hunk 8]
```diff
@@ -32,12 +32,7 @@ Spec(R::MPolyRing, I::MPolyIdeal)
 Spec(R::MPolyRing, U::AbsMPolyMultSet)
 Spec(R::MPolyRing, I::MPolyIdeal, U::AbsMPolyMultSet)
 ```
-
-### Copy constructors
-
-```@docs
-Spec(X::Spec)
-```
+See [`inclusion_morphism(::AbsSpec, ::AbsSpec)`](@ref) for a way to obtain the ideal ``I`` from ``X = \mathrm{Spec}(R, I)``.
```
Comment (fingolfin) [inline docs/src/AlgebraicGeometry/Schemes/AffineSchemes.md:35]: Maybe if you try this instead?
```suggestion
See [`inclusion_morphism(X::AbsSpec, Y::AbsSpec; check::Bool)`](@ref) for a way to obtain the ideal ``I`` from ``X = \mathrm{Spec}(R, I)``.
```
or perhaps
```suggestion
See [`inclusion_morphism(X::AbsSpec, Y::AbsSpec; check::Bool=true)`](@ref) for a way to obtain the ideal ``I`` from ``X = \mathrm{Spec}(R, I)``.
```

The other option would be to try and use a [Named doc `@ref`](https://documenter.juliadocs.org/stable/man/syntax/#Named-doc-@refs)

Finally, I would suggest to ask in the `#documentation` channel of the Julia Slack (not the OSCAR slack!) for hints on how to do this -- we might as well have hit on a bug in Documenter.jl here (wouldn't be the first time). The Documenter.jl documentation at times could be ... more precise and detailed and complete ... (just like ours ;-) )

**Comment by simonbrandhorst** [comment]
I leave the doc fix for another time.
